presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/clientproviders/TestRestryingPrestoThriftServiceProvider.java

```diff
@@ -0,0 +1,81 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.connector.thrift.clientproviders;
+
+import com.facebook.presto.connector.thrift.ThriftConnectorConfig;
+import com.facebook.presto.connector.thrift.api.PrestoThriftService;
+import com.facebook.presto.connector.thrift.api.PrestoThriftServiceException;
+import com.facebook.presto.connector.thrift.location.HostLocationProvider;
+import com.facebook.presto.connector.thrift.location.StaticLocationProvider;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import io.airlift.units.Duration;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.spi.HostAddress.fromParts;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+public class TestRestryingPrestoThriftServiceProvider
+{
+    private static final HostLocationProvider TEST_HOST_LOCATION_PROVIDER = new StaticLocationProvider(ImmutableList.of(fromParts("fakehost", 1234)));
+    private ListeningScheduledExecutorService retryExecutor;
+
+    @BeforeClass
+    void init()
+    {
+        retryExecutor = listeningDecorator(newScheduledThreadPool(2, threadsNamed("test-thrift-retry-driver-%s")));
+    }
+
+    @AfterClass
+    void shutdown()
+    {
+        try {
+            retryExecutor.shutdownNow();
+        }
+        catch (Exception ignore) {
+        }
+    }
+
+    @Test
+    void TestRetryService()
+            throws Exception
+    {
+        MockLocationAwarePrestoThriftServiceProvider mockServiceProvider = new MockLocationAwarePrestoThriftServiceProvider(TEST_HOST_LOCATION_PROVIDER);
+        ThriftConnectorConfig config = new ThriftConnectorConfig()
+                .setMaxRetryAttempts(3)
+                .setMinRetrySleepTime(new Duration(10, MILLISECONDS))
+                .setMaxRetrySleepTime(new Duration(20, MILLISECONDS));
+
+        RetryingPrestoThriftServiceProvider retryingServiceProvider =
+                new RetryingPrestoThriftServiceProvider(mockServiceProvider, retryExecutor, TEST_HOST_LOCATION_PROVIDER, config);
+
+        mockServiceProvider.setPendingFailuresToGenerate(2);
```

Comment:
Because the `MockLocationAwarePrestoThriftServiceProvider` is not stateless. It has configurable failures to be generated. We should have a provider per unit test.

Reply:
The question was about using a setter vs passing it in constructor.